### PR TITLE
fix(connector): fix batch file source when there are no files

### LINKED ISSUE #17984: batch select from s3 source hangs
### Describe the bug

Reported in community
https://risingwave-community.slack.com/archives/C03BW71523T/p1723121143187519

### Error message/log

_No response_

### To Reproduce

Docker compose:
```yml
services:
  minio:
    image: minio/minio:latest
    container_name: minio
    environment:
      MINIO_ROOT_USER: mlops
      MINIO_ROOT_PASSWORD: password
    entrypoint: "
      /bin/sh -c '

      set -e

      mkdir -p \"/data/hummock001\"
      mkdir -p \"/data/monitoring\" 

      /usr/bin/docker-entrypoint.sh \"$$0\" \"$$@\"

      '"
    command: server /data --console-address ":9001"
    ports:
      - "9000:9000"
      - "9001:9001"
      - "38223:38223"
    volumes:
      - minio_data:/data
    healthcheck:
      test: ["CMD", "mc", "ready", "local"]
      interval: 5s
      timeout: 5s
      retries: 5

  postgres:
    image: "postgres:15-alpine"
    environment:
      - POSTGRES_HOST_AUTH_METHOD=trust
      - POSTGRES_USER=postgres
      - POSTGRES_DB=metadata
      - POSTGRES_INITDB_ARGS=--encoding=UTF-8 --lc-collate=C --lc-ctype=C
    expose:
      - "5432"
    ports:
      - "8432:5432"
    volumes:
      - "postgres:/var/lib/postgresql/data"
    healthcheck:
      test: [ "CMD-SHELL", "pg_isready -U postgres" ]
      interval: 2s
      timeout: 5s
      retries: 5
    restart: always

  risingwave-standalone:
    image: risingwavelabs/risingwave:latest
    container_name: risingwave-standalone
    command: "standalone --meta-opts=\" \
                    --listen-addr 0.0.0.0:5690 \
                    --advertise-addr 0.0.0.0:5690 \
                    --dashboard-host 0.0.0.0:5691 \
                    --prometheus-host 0.0.0.0:1250 \
                    --backend sql \
                    --sql-endpoint postgres://postgres:@postgres:5432/metadata \
                    --state-store hummock+minio://mlops:password@minio:9000/hummock001 \
                    --data-directory hummock_001 \
                    --config-path /risingwave.toml\" \
                 --compute-opts=\" \
                    --config-path /risingwave.toml \
                    --listen-addr 0.0.0.0:5688 \
                    --prometheus-listener-addr 0.0.0.0:1250 \
                    --advertise-addr 0.0.0.0:5688 \
                    --async-stack-trace verbose \
                    #--parallelism 4 \
                    #--total-memory-bytes 8589934592 \
                    --role both \
                    --meta-address http://0.0.0.0:5690/\" \
                 --frontend-opts=\" \
                   --config-path /risingwave.toml \
                   --listen-addr 0.0.0.0:4566 \
                   --advertise-addr 0.0.0.0:4566 \
                   --prometheus-listener-addr 0.0.0.0:1250 \
                   --health-check-listener-addr 0.0.0.0:6786 \
                   --meta-addr http://0.0.0.0:5690/\" \
                 --compactor-opts=\" \
                   --listen-addr 0.0.0.0:6660 \
                   --prometheus-listener-addr 0.0.0.0:1250 \
                   --advertise-addr 0.0.0.0:6660 \
                   --meta-address http://0.0.0.0:5690/\""
    expose:
      - "6660"
      - "4566"
      - "5688"
      - "5690"
      - "1250"
      - "5691"
    ports:
      - "4566:4566"
      - "5690:5690"
      - "5691:5691"
      - "1250:1250"
    depends_on:
      - postgres
      - minio
    volumes:
      - "./risingwave.toml:/risingwave.toml"
    environment:
      RUST_BACKTRACE: "1"
      # If ENABLE_TELEMETRY is not set, telemetry will start by default
      ENABLE_TELEMETRY: ${ENABLE_TELEMETRY:-true}
      RW_TELEMETRY_TYPE: ${RW_TELEMETRY_TYPE:-"docker-compose"}
      RW_SECRET_STORE_PRIVATE_KEY_HEX: ${RW_SECRET_STORE_PRIVATE_KEY_HEX:-0123456789abcdef}
    healthcheck:
      test:
        - CMD-SHELL
        - bash -c 'printf \"GET / HTTP/1.1\n\n\" > /dev/tcp/127.0.0.1/6660; exit $$?;'
        - bash -c 'printf \"GET / HTTP/1.1\n\n\" > /dev/tcp/127.0.0.1/5688; exit $$?;'
        - bash -c 'printf \"GET / HTTP/1.1\n\n\" > /dev/tcp/127.0.0.1/4566; exit $$?;'
        - bash -c 'printf \"GET / HTTP/1.1\n\n\" > /dev/tcp/127.0.0.1/5690; exit $$?;'
      interval: 1s
      timeout: 5s
    restart: always
    deploy:
      resources:
        limits:
          memory: 28G
        reservations:
          memory: 28G

  superset:
    build:
      context: ./superset
      dockerfile: dockerfile
    container_name: superset
    environment:
      - ADMIN_USERNAME=admin
      - ADMIN_EMAIL=admin@superset.com
      - ADMIN_PASSWORD=password
    ports:
      - '19088:8088'
    volumes:
      - superset_home:/app/superset_home

volumes:
  minio_data:
    external: false
  postgres:
    external: false
  superset_home:
    external: false
```

`example.json`:
```
{"id":1,"name":"Adair","amount":75.48,"operation_time":"2024-05-04T01:16:02Z"}
```

```sql
CREATE
SOURCE minio_expanses (
                   id int ,
                   name TEXT,
                   amount NUMERIC,
                   operation_time TIMESTAMP WITH TIME ZONE,
) WITH ( connector = 's3_v2',
    s3.region_name = 'us-east-1',
    s3.bucket_name = 'monitoring',
    s3.credentials.access = '<ACCESS_KEY>',
    s3.credentials.secret = '<SECRET>',
    s3.endpoint_url = 'http://minio:9000/') FORMAT PLAIN ENCODE JSON;

SELECT * FROM minio_expanses;
```

### Expected behavior

_No response_

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Previously for batch read from file source, the task_parallelism is set to `(self.batch_parallelism / 2) as u32`, which by default is 2, and this will cause 2 issues:
- when their is only one file, another parallelism can not cancel normally.
- when their is no files in source, the  parallelism can not cancel.

This pr fix these two issues by
1. set the `task_parallelism` to `min(file_nums, (batch_parallelism / 2))`, and minimum value is 1.
2. When their is no files in source, schedule an empty task to prevent stuck.

close #17984

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by tabVersion — APPROVED** [review]
the fix LGTM, may consider adding a test?

**Review by xxchan — COMMENTED** [review]
fix https://github.com/risingwavelabs/risingwave/issues/17984?

**Comment by wcy-fdu** [comment]
> fix #17984?

Yes, added in pr description.

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 1]
```diff
@@ -380,9 +380,33 @@ impl StageRunner {
...
-            let chunk_size = (source_info.split_info().unwrap().len() as f32
+            // If their is no file in source, the `chunk_size` is set to 1.
+            let chunk_size = ((source_info.split_info().unwrap().len() as f32
                 / self.stage.parallelism.unwrap() as f32)
-                .ceil() as usize;
+                .ceil() as usize)
+                .max(1);
+            // No file in source, schedule an empty task.
+            if source_info.split_info().unwrap().is_empty() {
+                const EMPTY_TASK_ID: u64 = 0;
```
Comment (chenzl25) [inline src/frontend/src/scheduler/distributed/stage.rs:390]: Can we change it into this structure:
```
if source_info.split_info().unwrap().is_empty() {
   // empty task
} else {
  let chunk_size = balala.
  // rest 
}
```

**Review by chenzl25 — COMMENTED** [review]
Rest LGTM

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 2]
```diff
@@ -380,30 +380,57 @@ impl StageRunner {
                 ));
             }
         } else if let Some(source_info) = self.stage.source_info.as_ref() {
-            let chunk_size = (source_info.split_info().unwrap().len() as f32
+            // If their is no file in source, the `chunk_size` is set to 1.
```
Comment (KeXiangWang) [inline src/frontend/src/scheduler/distributed/stage.rs:383]:
```suggestion
            // If there is no file in source, the `chunk_size` is set to 1.
```

**Review by KeXiangWang — APPROVED** [review]

**Pull request merged**